### PR TITLE
Initial version of cc_fuzz_run, which creates a script <name>_run

### LINKED ISSUE #18: Augment the cc_fuzz_test() macro with a new <name>_run rule
The _run rule should invoke the fuzzer runner tool (#17) when run using "bazel run ...".

### PR DESCRIPTION
**Pull request by tengpengli**
**Commit Message:**
Initial version of cc_fuzz_run, which creates a script \<name\>_run
The <name>_run calls the launcher script to run the \<name\> executable
**Additional Description:**
The only difference between this PR and #31 is the cc_deps.bzl.
The command inside the generated script is absolutely not what we are looking for, but without timeout, the bazel run never exists. How can I make the script end? And how to pass the timeout_secs to the launcher script?
**Testing:**
CI testing and local testing.
**Docs Changes:** N/A
**Fixes #18**

### CODE REVIEW

File: fuzzing/cc_deps.bzl [hunk 1]
```diff
@@ -17,6 +17,50 @@
...
+    )
+    ctx.actions.write(script, script_content, is_executable = True)
+
+    # Merge the two dependencies
+    runfiles = ctx.attr._launcher[DefaultInfo].default_runfiles
+    runfiles = runfiles.merge(ctx.attr.dep[DefaultInfo].default_runfiles)
+
+    return [DefaultInfo(executable = script, runfiles = runfiles)]
+
+# What should the argument be? How to pass timeout_secs?
```
Comment (stefanbucur) [inline fuzzing/cc_deps.bzl:45]: Re. the `--timeout_secs` question, wouldn't something like this work?

```
$ bazel run --config=asan-libfuzzer //path/to:fuzzer_run -- --timeout_secs=60
```

File: fuzzing/cc_deps.bzl [hunk 2]
```diff
@@ -17,6 +17,50 @@
...
+        # The launcher script to start fuzzing test
+        "_launcher": attr.label(
+            default = Label("//fuzzing/tools:launcher"),
+            executable = True,
+            cfg = "host",
+        ),
+        # The _fuzz_test executable to run
+        "dep": attr.label(
+            executable = True,
+            cfg = "host",
```
Comment (stefanbucur) [inline fuzzing/cc_deps.bzl:58]: Should this be "target" instead? I guess we want the `asan-libfuzzer` configuration and all other build settings to apply on this one.

File: fuzzing/cc_deps.bzl [hunk 3]
```diff
@@ -17,6 +17,50 @@
...
+    runfiles = ctx.attr._launcher[DefaultInfo].default_runfiles
+    runfiles = runfiles.merge(ctx.attr.dep[DefaultInfo].default_runfiles)
+
+    return [DefaultInfo(executable = script, runfiles = runfiles)]
+
+# What should the argument be? How to pass timeout_secs?
+cc_fuzz_run = rule(
+    implementation = _cc_fuzz_run_impl,
+    attrs = {
+        # The launcher script to start fuzzing test
```
Comment (stefanbucur) [inline fuzzing/cc_deps.bzl:49]: .bzl source code comments will not be visible to developers using this rule. Consider turning them into `doc` fields for each attribute defined. This is documented in more depth here: https://github.com/bazelbuild/stardoc/blob/master/docs/writing_stardoc.md

File: fuzzing/cc_deps.bzl [hunk 4]
```diff
@@ -17,6 +17,50 @@
...
+    ctx.actions.write(script, script_content, is_executable = True)
+
+    # Merge the two dependencies
+    runfiles = ctx.attr._launcher[DefaultInfo].default_runfiles
+    runfiles = runfiles.merge(ctx.attr.dep[DefaultInfo].default_runfiles)
+
+    return [DefaultInfo(executable = script, runfiles = runfiles)]
+
+# What should the argument be? How to pass timeout_secs?
+cc_fuzz_run = rule(
```
Comment (stefanbucur) [inline fuzzing/cc_deps.bzl:46]: I imagine this launcher is not specific to C++ fuzz targets, so perhaps call this rule something more generic? I'd suggest `fuzzing_launcher` (and similarly, `_fuzzing_launcher_impl` for the rule implementation).

File: fuzzing/cc_deps.bzl (hunk 1 above)
Comment (asraa) [reply fuzzing/cc_deps.bzl:45]: Agreed, passing args through the CLI probably works in this case., and mimics normal fuzzing invocations. 

I think you could also add an explicit arg for a timeout to the rule, or maybe other ways, but the above usage would be good (and I think should work).

File: fuzzing/cc_deps.bzl (hunk 4 above)
Comment (tengpengli) [reply fuzzing/cc_deps.bzl:46]: With that name, do you think it will be more proper to place this rule into common.bzl?

Comment (stefanbucur) [reply fuzzing/cc_deps.bzl:46]: Good point, yes, placing it in common.bzl makes a lot of sense.

**Comment by tengpengli** [comment]
I think I found the reason why my <name>_run script never ends.

In my local test, I was using empty_fuzz_test_run to test the script. The `--timeout_secs` argument at the very end is passed as the libfuzzer argument `-timeout`, which set the time limit for **one round of fuzz test**. In the meantime, the empty_fuzz_test  function simply returns, which ends really quickly, thus the timeout would never be triggered. In other words, the script never ends.

**Comment by stefanbucur** [comment]
You're absolutely right! It was the wrong flag to use, didn't even notice that :) It seems like https://llvm.org/docs/LibFuzzer.html has also a `-max_total_time` flag: can you try that and see if it works?

**Comment by tengpengli** [comment]
> You're absolutely right! It was the wrong flag to use, didn't even notice that :) It seems like https://llvm.org/docs/LibFuzzer.html has also a `-max_total_time` flag: can you try that and see if it works?

It works for empty_fuzz_run, but the interesting thing is that the oom_fuzz_test and hang_fuzz_test don't get interrupted. 

I only tried those three, maybe it's because the `-max_total_time` doesn't set a timer for the child thread created by the fuzzing test, but only a timer on the parent thread? 

Anyway, I use both `-timeout` and `-max_total_time`, now it runs as expected.

**Comment by tengpengli** [comment]
PTAL!

For the timeout problem in the launcher script, I fixed it in #35

File: fuzzing/cc_deps.bzl [hunk 5]
```diff
@@ -13,9 +13,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-""" This file contains basic functions for fuzz test. """
+""" This file contains basic functions for cc fuzz test. """
```
Comment (stefanbucur) [inline fuzzing/cc_deps.bzl:16]: Nit: Remove the extra space between the quotes and the text?

**Pull request merged**